core/src/main/java/io/questdb/cairo/mv/MatViewRefreshJob.java

```diff
@@ -178,62 +180,152 @@ private void enqueueInvalidateDependentViews(TableToken viewToken, String invali
+            // Full refresh, i.e. someone has run REFRESH FULL SQL.
+            // When the table is empty, min/max timestamps are set to Long.MAX_VALUE / LONG.MIN_VALUE,
+            // so we end up skipping the refresh.
             minTs = baseTableReader.getMinTimestamp();
             maxTs = baseTableReader.getMaxTimestamp();
         }
 
+        // In case of incremental or full refresh we may need to include complete periods
+        // and/or remove intervals older than the refresh limit.
+        if (!rangeRefresh) {
+            // Check if we're doing incremental/full refresh on a period mat view.
+            // If so, we may need to remove incomplete periods from the final refresh interval.
+            if (viewDefinition.getPeriodLength() > 0) {
+                TimestampSampler periodSampler = viewDefinition.getPeriodSampler();
+                if (periodSampler == null) {
+                    periodSampler = TimestampSamplerFactory.getInstance(viewDefinition.getPeriodLength(), viewDefinition.getPeriodLengthUnit(), -1);
+                    viewDefinition.setPeriodSampler(periodSampler);
+                }
+                periodSampler.setStart(viewDefinition.getTimerStart());
+
+                final long delay = MatViewTimerJob.periodDelayMicros(viewDefinition.getPeriodDelay(), viewDefinition.getPeriodDelayUnit());
+                long nowLocal = viewDefinition.getTimerTzRules() != null
+                        ? now + viewDefinition.getTimerTzRules().getOffset(now)
+                        : now;
+                // Period hi is exclusive, but maxTs is inclusive, so we align them.
+                final long periodHiLocal = periodSampler.round(nowLocal - delay) - 1;
+                final long periodHi = viewDefinition.getTimerTzRules() != null
+                        ? periodHiLocal - viewDefinition.getTimerTzRules().getOffset(periodHiLocal)
+                        : periodHiLocal;
+
+                // Remove the incomplete period from both txn intervals and refresh interval.
+                if (txnIntervals != null && txnIntervals.size() > 0) {
+                    txnIntervals.add(Long.MIN_VALUE, periodHi);
+                    IntervalUtils.intersectInPlace(txnIntervals, txnIntervals.size() - 2);
+                }
+                maxTs = Math.min(maxTs, periodHi);
+
+                if (incrementalRefresh) {
+                    // Incremental refresh on a period mat view works in a special way.
+                    // We need to refresh whatever is in the txn intervals and all periods
+                    // that became complete since the last refresh.
+                    long periodLo = viewState.getLastPeriodHi();
+                    if (periodLo == Numbers.LONG_NULL) {
+                        periodLo = baseTableReader.getMinTimestamp();
+                    }
+                    if (periodLo < periodHi) {
+                        txnIntervals.add(periodLo, periodHi);
```

Comment:
nit: this can be a reusable method:

```
txnIntervals.add(periodLo, periodHi);
IntervalUtils.unionInPlace(txnIntervals, txnIntervals.size() - 2);
```